examples/minimal/packages/client-react-external-wallet/src/mud/customClient.ts

```diff
@@ -0,0 +1,103 @@
+import { type Subject } from "rxjs";
+import { getNonceManager, type ContractWrite } from "@latticexyz/common";
+import { type SyncResult } from "@latticexyz/store-sync";
+
+const burnerBlockTag = "pending";
+
+// See @latticexyz/common/src/sendTransaction.ts
+export const burnerActions = <transport extends Transport, chain extends Chain, account extends Account>(
+  // The client must be extended with wallet and public actions.
+  // https://viem.sh/docs/accounts/local#optional-extend-with-public-actions
+  client: Client<
+    transport,
+    chain,
+    account,
+    WalletRpcSchema & PublicRpcSchema,
+    WalletActions<chain, account> & PublicActions<transport, chain>
+  >
+): Pick<WalletActions<chain, account>, "sendTransaction"> &
+  Pick<PublicActions<transport, chain>, "simulateContract"> => {
+  // TODO: Use the `debug` library once this function has been moved to the `common` library.
+  const debug = console.log;
+
+  return {
+    // Applies to: `client.(sendTransaction|writeContract)`, `getContract(client, ...).write`
+    sendTransaction: async (args) => {
+      const nonceManager = await getNonceManager({
+        client,
+        address: client.account.address,
+        blockTag: burnerBlockTag,
+      });
+
+      return nonceManager.mempoolQueue.add(
+        () =>
+          pRetry(
+            async () => {
+              if (!nonceManager.hasNonce()) {
+                await nonceManager.resetNonce();
+              }
+
+              const nonce = nonceManager.nextNonce();
+              debug("sending tx with nonce", nonce, "to", args.to);
+              return sendTransaction(client, { ...args, nonce } as typeof args);
+            },
+            {
+              retries: 3,
+              onFailedAttempt: async (error) => {
+                // On nonce errors, reset the nonce and retry
+                if (nonceManager.shouldResetNonce(error)) {
+                  debug("got nonce error, retrying", error.message);
+                  await nonceManager.resetNonce();
+                  return;
+                }
+                // TODO: prepare again if there are gas errors?
+                throw error;
+              },
+            }
+          ),
+        { throwOnTimeout: true }
+      );
+    },
+    // Applies to: `client.simulateContract`, `getContract(client, ...).simulate`
+    simulateContract: (args) => simulateContract(client, { ...args, blockTag: burnerBlockTag } as typeof args),
+  };
+};
+
+// See @latticexyz/common/src/getContract.ts
+export const setupObserverActions = (write$: Subject<ContractWrite>) => {
+  let nextWriteId = 0;
+
+  return <transport extends Transport, chain extends Chain, account extends Account>(
+    client: WalletClient<transport, chain, account>
+  ): Pick<WalletActions<chain, account>, "writeContract"> => ({
+    // Applies to: `client.writeContract`, `getContract(client, ...).write`
+    writeContract: async (args) => {
+      const result = writeContract(client, args);
```

Comment:
It turned out I couldn't use `getAction` here. This is because if I use `getAction(client, writeContract, "writeContract")(args)`, it uses the normal `walletActions`' `writeContract`, which in turn uses the normal `sendTransaction`. As a result, `burnerActions`' `sendTransaction` isn't invoked. When extending a client, the `client` argument in the action function is captured at that point, making `getAction` unsuitable for this use case.

https://github.com/wevm/viem/blob/viem%402.7.12/src/clients/decorators/wallet.ts#L684-L710

Reply:
internally, viem's `writeContract` does call `getAction('sendTransaction')`:
https://github.com/wevm/viem/blob/main/src/actions/wallet/writeContract.ts#L163-L167

so I am wondering if I am misunderstanding the intent

Reply:
Thanks for the comment! I realized that we first need to separate the custom client to move forward, so I’ll address this comment in the independent custom client PR, which I’ll be creating shortly.